<?xml version='1.0'?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="train3">
 <xacro:property name="pi2" value="1.57079632679" />
 <xacro:property name="body_length" value="3" />
 <xacro:property name="axel_offset" value="0.5" />
 <xacro:property name="axel_offset2" value="-0.5" />
 <xacro:property name="axel_z_offset" value="-0.3" />




 <xacro:macro name="wheel_bumper_macro" params="reference">
 <gazebo reference="${reference}">
   <!-- contact sensor -->
   <sensor type="contact" name="${reference}_contact_sensor">
     <update_rate>50.0</update_rate>
     <always_on>true</always_on>
     <contact>
       <collision>${reference}_collision</collision>
     </contact>
     <plugin name="lside_bumper_plugin" filename="libgazebo_ros_bumper.so">
       <bumperTopicName>bumper_${reference}</bumperTopicName>
       <frameName>World</frameName>
  </plugin>
   </sensor>
 </gazebo>
 </xacro:macro>



  <xacro:macro name="wheel_macro" params="name model">
    <link name="${name}">
     <inertial>
      <mass value="7.5"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.49" ixy="0" ixz="0" iyy="0.37" iyz="0" izz="0.37"/>
     </inertial>
     <collision>
       <origin xyz="0 0 0" rpy="0 1.57079632679 0"/>
       <laser_retro>0</laser_retro>
       <max_contacts>10</max_contacts>
       <pose >0 0 0 0 0 0</pose>
       <geometry>
         <cylinder length="0.45" radius="0.28"/>
       </geometry>
     </collision>
     <visual name="visual">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
       <mesh filename="package://ros_robotics/meshes/${model}" scale="1 1 1"/>
      </geometry>
      <material name="silver">
       <color rgba="0.75 0.75 0.75 1"/>
      </material>
     </visual>
    </link>
    <xacro:wheel_bumper_macro reference="${name}"/>
  </xacro:macro>


 <xacro:macro name="pivot_joint_macro" params="xyz name parent child">
  <joint name="${name}" type="revolute">
   <parent link="${parent}"/>
   <child link="${child}"/>
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <axis xyz="0  0  1"/>
   <limit effort="-1" lower="-0.18" upper="0.18" velocity="-1"/>
   <!--limit effort="-1" lower="-0.001" upper="0.001" velocity="-1"/-->
  </joint>
  <xacro:trans_macro name="${name}"/>
 </xacro:macro>

<xacro:macro name="body_macro" params="xyz name">
 <link name="${name}">
  <inertial>
   <origin xyz="${xyz}"/>
   <mass value="15"/>
   <inertia ixx="0.166667" ixy="0" ixz="0" iyy="0.166667" iyz="0" izz="0.166667"/>
  </inertial>
  <collision name="collision">
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <geometry>
     <box size="2 0.5 0.5" />
   </geometry>
  </collision>
  <visual name="visual">
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <geometry>
     <box size="2 0.5 0.5" />
   </geometry>
   <material name="silver">
    <color rgba="0.75 0.75 0.75 1"/>
   </material>
  </visual>
  </link>
 </xacro:macro>

 <xacro:macro name="trans_macro" params="name">
  <transmission name="simple_trans">
   <type>transmission_interface/SimpleTransmission</type>
   <joint name="${name}">
     <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
   </joint>
   <actuator name="motor_${name}">
     <mechanicalReduction>1</mechanicalReduction>
   </actuator>
  </transmission>
 </xacro:macro>

<xacro:macro name="axle_macro" params="xyz name">
 <link name="${name}">
  <inertial>
   <mass value="30"/>
   <origin rpy="0  0  0" xyz="0  0  0"/>
   <inertia ixx="0.166667" ixy="0" ixz="0" iyy="0.166667" iyz="0" izz="0.166667"/>
  </inertial>
  <collision name="collision">
   <origin rpy="0  0  0" xyz="0    0   0"/>
   <geometry>
    <mesh filename="package://ros_robotics/meshes/Front_Axle.STL" scale="1.0 1.0 1.0"/>
   </geometry>
  </collision>
  <visual name="visual">
   <origin rpy="0  0  0" xyz="0 0 0"/>
   <geometry>
    <mesh filename="package://ros_robotics/meshes/Front_Axle.STL" scale="1 1 1"/>
   </geometry>
   <material name="silver">
    <color rgba="0.75 0.75 0.75 1"/>
   </material>
  </visual>
 </link>
</xacro:macro>



<xacro:macro name="joint_wheel_macro" params="xyz name parent child">
 <joint name="${name}" type="revolute">
  <parent link="${parent}"/>
  <child link="${child}"/>
  <origin rpy="0    0    1.57079632679" xyz="${xyz}"/>
  <axis xyz="1  0  0"/>
  <limit effort="200" lower="-1e+16" upper="1e+16" velocity="-1"/>
 </joint>
 <xacro:trans_macro name="${name}"/>
</xacro:macro>

 <xacro:macro name="body_joint_macro" params="xyz axis name parent child">
  <joint name="${name}" type="revolute">
   <parent link="${parent}"/>
   <child link="${child}"/>
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <axis xyz="${axis}"/>
   <limit effort="-1" lower="-1.57" upper="1.57" velocity="-1"/>
  </joint>
   <xacro:trans_macro name="${name}"/>
 </xacro:macro>

 <xacro:macro name="fixed_joint_macro" params="xyz name parent child">
  <joint name="${name}" type="prismatic">
   <parent link="${parent}"/>
   <child link="${child}"/>
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <axis xyz="0 0 1"/>
   <limit effort="-1" lower="-0.01" upper="0.01" velocity="-1"/>
  </joint>
 </xacro:macro>

<xacro:macro name="dummy_link_macro" params="xyz name">
 <link name="${name}">
  <inertial>
   <mass value="0.001"/>
   <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
  </inertial>
  <visual name="visual">
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <geometry>
     <sphere radius="0.25"/>
   </geometry>
   <material name="silver">
    <color rgba="0.75 0.75 0.75 1"/>
   </material>
  </visual>
  </link>
 </xacro:macro>

<!-- body one -->
<xacro:body_macro name="body_1" xyz="0 0 0"/>

<xacro:pivot_joint_macro name="pivot_joint_1_1" parent="body_1" child="axle_1_1" xyz="${axel_offset} 0 ${axel_z_offset}"/>
<xacro:axle_macro name="axle_1_1" xyz="0 0 0"/>
<xacro:joint_wheel_macro xyz="0  0.57526 0" name="joint_wheel_1_1_1" parent="axle_1_1" child="wheel_1_1_1"/>
<xacro:joint_wheel_macro xyz="0  -0.57526 0" name="joint_wheel_1_1_2" parent="axle_1_1" child="wheel_1_1_2"/>
<xacro:wheel_macro name="wheel_1_1_1" model="front_left_wheel.STL"/>
<xacro:wheel_macro name="wheel_1_1_2" model="front_left_wheel.STL"/>

<xacro:pivot_joint_macro name="pivot_joint_1_2" parent="body_1" child="axle_1_2" xyz="${axel_offset2} 0 ${axel_z_offset}"/>
<xacro:axle_macro name="axle_1_2" xyz="0 0 0"/>
<xacro:joint_wheel_macro xyz="0  0.57526 0" name="joint_wheel_1_2_1" parent="axle_1_2" child="wheel_1_2_1"/>
<xacro:joint_wheel_macro xyz="0  -0.57526 0" name="joint_wheel_1_2_2" parent="axle_1_2" child="wheel_1_2_2"/>
<xacro:wheel_macro name="wheel_1_2_1" model="front_left_wheel.STL"/>
<xacro:wheel_macro name="wheel_1_2_2" model="front_left_wheel.STL"/>

<!-- 1->2 link -->
<xacro:body_joint_macro name="body_joint_1_z" parent="body_1" child="dummy_1_1" xyz="${-body_length/2} 0 0" axis="0 0 1"/>
<xacro:dummy_link_macro name="dummy_1_1" xyz="0 0 0"/>
<xacro:body_joint_macro name="body_joint_1_y" parent="dummy_1_1" child="dummy_1_2" xyz="0 0 0" axis="0 1 0"/>
<xacro:dummy_link_macro name="dummy_1_2" xyz="0 0 0"/>
<xacro:body_joint_macro name="body_joint_1_x" parent="dummy_1_2" child="dummy_1_3" xyz="0 0 0" axis="1 0 0"/>
<xacro:dummy_link_macro name="dummy_1_3" xyz="0 0 0"/>


<xacro:fixed_joint_macro name="fixed_joint_1" parent="dummy_1_3" child="body_2" xyz="${-body_length/2} 0 0"/>

<!-- body two -->
<xacro:body_macro name="body_2" xyz="0 0 0"/>
<xacro:pivot_joint_macro name="pivot_joint_2_1" parent="body_2" child="axle_2_1" xyz="${axel_offset} 0 ${axel_z_offset}"/>
<xacro:axle_macro name="axle_2_1" xyz="0 0 0"/>
<xacro:joint_wheel_macro xyz="0  0.57526 0" name="joint_wheel_2_1_1" parent="axle_2_1" child="wheel_2_1_1"/>
<xacro:joint_wheel_macro xyz="0  -0.57526 0" name="joint_wheel_2_1_2" parent="axle_2_1" child="wheel_2_1_2"/>
<xacro:wheel_macro name="wheel_2_1_1" model="front_left_wheel.STL"/>
<xacro:wheel_macro name="wheel_2_1_2" model="front_left_wheel.STL"/>

<xacro:pivot_joint_macro name="pivot_joint_2_2" parent="body_2" child="axle_2_2" xyz="${axel_offset2} 0 ${axel_z_offset}"/>
<xacro:axle_macro name="axle_2_2" xyz="0 0 0"/>
<xacro:joint_wheel_macro xyz="0  0.57526 0" name="joint_wheel_2_2_1" parent="axle_2_2" child="wheel_2_2_1"/>
<xacro:joint_wheel_macro xyz="0  -0.57526 0" name="joint_wheel_2_2_2" parent="axle_2_2" child="wheel_2_2_2"/>
<xacro:wheel_macro name="wheel_2_2_1" model="front_left_wheel.STL"/>
<xacro:wheel_macro name="wheel_2_2_2" model="front_left_wheel.STL"/>

<!-- 2->3 link -->
<xacro:body_joint_macro name="body_joint_2_z" parent="body_2" child="dummy_2_1" xyz="${-body_length/2} 0 0" axis="0 0 1"/>
<xacro:dummy_link_macro name="dummy_2_1" xyz="0 0 0"/>
<xacro:body_joint_macro name="body_joint_2_y" parent="dummy_2_1" child="dummy_2_2" xyz="0 0 0" axis="0 1 0"/>
<xacro:dummy_link_macro name="dummy_2_2" xyz="0 0 0"/>
<xacro:body_joint_macro name="body_joint_2_x" parent="dummy_2_2" child="dummy_2_3" xyz="0 0 0" axis="1 0 0"/>
<xacro:dummy_link_macro name="dummy_2_3" xyz="0 0 0"/>

<xacro:fixed_joint_macro name="fixed_joint_2" parent="dummy_2_3" child="body_3" xyz="${-body_length/2} 0 0"/>

<!-- body three -->

<xacro:body_macro name="body_3" xyz="0 0 0"/>

<xacro:pivot_joint_macro name="pivot_joint_3_1" parent="body_3" child="axle_3_1" xyz="${axel_offset} 0 ${axel_z_offset}"/>
<xacro:axle_macro name="axle_3_1" xyz="0 0 0"/>
<xacro:joint_wheel_macro xyz="0  0.57526 0" name="joint_wheel_3_1_1" parent="axle_3_1" child="wheel_3_1_1"/>
<xacro:joint_wheel_macro xyz="0  -0.57526 0" name="joint_wheel_3_1_2" parent="axle_3_1" child="wheel_3_1_2"/>
<xacro:wheel_macro name="wheel_3_1_1" model="front_left_wheel.STL"/>
<xacro:wheel_macro name="wheel_3_1_2" model="front_left_wheel.STL"/>

<xacro:pivot_joint_macro name="pivot_joint_3_2" parent="body_3" child="axle_3_2" xyz="${axel_offset2} 0 ${axel_z_offset}"/>
<xacro:axle_macro name="axle_3_2" xyz="0 0 0"/>
<xacro:joint_wheel_macro xyz="0  0.57526 0" name="joint_wheel_3_2_1" parent="axle_3_2" child="wheel_3_2_1"/>
<xacro:joint_wheel_macro xyz="0  -0.57526 0" name="joint_wheel_3_2_2" parent="axle_3_2" child="wheel_3_2_2"/>
<xacro:wheel_macro name="wheel_3_2_1" model="front_left_wheel.STL"/>
<xacro:wheel_macro name="wheel_3_2_2" model="front_left_wheel.STL"/>

<!-- https://answers.ros.org/question/258420/ball-joint-in-urdf/ -->

<gazebo>
  <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
   <robotNamespace>/dd_robot</robotNamespace>
   <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
   <legacyModeNS>true</legacyModeNS>
  </plugin>
</gazebo>


</robot>
